mlflow/bedrock/stream.py

```diff
@@ -0,0 +1,167 @@
+
+    Args:
+        stream: The original event stream to wrap.
+        client: The MLflow client to end the span.
+        span: The span to record events and response in.
+        inputs: The inputs to the converse API.
+    """
+
+    def __init__(
+        self,
+        stream: EventStream,
+        client: MlflowClient,
+        span: LiveSpan,
+        inputs: Optional[dict[str, Any]] = None,
+    ):
+        self._stream = stream
+        self._span = span
+        self._client = client
+        self._inputs = inputs
+
+    def __iter__(self):
+        for event in self._stream:
+            self._handle_event(self._span, event)
+            yield event
+
+        # End the span when the stream is exhausted
+        self._close()
+
+    def __getattr__(self, attr):
+        """Delegate all other attributes to the original stream."""
+        return getattr(self._stream, attr)
+
+    def _handle_event(self, span, event):
+        """Process a single event from the stream."""
+        raise NotImplementedError
+
+    def _close(self):
+        """End the span and run any finalization logic."""
+        raise NotImplementedError
+
+    @capture_exception("Failed to handle event for the stream")
+    def _end_span(self):
+        """End the span."""
+        if self._span.parent_id:
+            self._client.end_span(self._span.request_id, self._span.span_id)
+        else:
+            self._client.end_trace(self._span.request_id)
+
+
+class InvokeModelStreamWrapper(BaseEventStreamWrapper):
+    """A wrapper class for a event stream returned by the InvokeModelWithResponseStream API."""
+
+    @capture_exception("Failed to handle event for the stream")
+    def _handle_event(self, span, event):
+        chunk = json.loads(event["chunk"]["bytes"])
+        self._span.add_event(SpanEvent(name=chunk["type"], attributes={"json": json.dumps(chunk)}))
+
+    def _close(self):
+        self._end_span()
+
+
+class ConverseStreamWrapper(BaseEventStreamWrapper):
+    """A wrapper class for a event stream returned by the ConverseStream API."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._response_builder = _ConverseMessageBuilder()
+
+    def __getattr__(self, attr):
+        """Delegate all other attributes to the original stream."""
+        return getattr(self._stream, attr)
+
+    @capture_exception("Failed to handle event for the stream")
+    def _handle_event(self, span, event):
+        """
+        Process a single event from the stream.
+
+        Refer to the following documentation for the event format:
+        https://boto3.amazonaws.com/v1/documentation/api/1.35.8/reference/services/bedrock-runtime/client/converse_stream.html
+        """
+        event_name = list(event.keys())[0]
+        self._response_builder.process_event(event_name, event[event_name])
+        # Record raw event as a span event
+        self._span.add_event(
```

Comment:
I think this is the right approach. I don't imagine the raw stream chunk data to be accessed frequently (the debuggability of things like early / unexpected termination is probably so infrequent that having this in any other collection location is pointless). +1